File: LeetCode/maximum-sum-circular-subarray.cpp
// https://leetcode.com/problems/maximum-sum-circular-subarray/

// Enumerate prefix and suffix sums
/*
Time Complexity => O(N)
Space Complexity => O(N)
*/

class Solution
{
public:
	int maxSubarraySumCircular(vector<int> &nums)
	{
		const int n = nums.size();
		vector<int> right_max(n);
		right_max[n - 1] = nums[n - 1];
		for (int suffix_sum = nums[n - 1], i = n - 2; i >= 0; --i)
		{
			suffix_sum += nums[i];
			right_max[i] = max(right_max[i + 1], suffix_sum);
		}
		int max_sum = nums[0];
		int special_sum = nums[0];
		for (int i = 0, suffix_sum = 0, curMax = 0; i < n; ++i)
		{
			curMax = max(curMax, 0) + nums[i];
			// This is Kadane's algorithm.
			max_sum = max(max_sum, curMax);
			suffix_sum += nums[i];
			if (i + 1 < n)
			{
				special_sum = max(special_sum, suffix_sum + right_max[i + 1]);
			}
		}
		return max(max_sum, special_sum);
	}
};

// Calculate the "Minimum Subarray"
/*
Time Complexity => O(N)
Space Complexity => O(1)
*/

class Solution
{
public:
	int maxSubarraySumCircular(vector<int> &nums)
	{
		int cur_max = 0, cur_min = 0, sum = 0, max_sum = nums[0], min_sum = nums[0];
		for (int num : nums)
		{
			cur_max = max(cur_max, 0) + num;
			max_sum = max(max_sum, cur_max);
			cur_min = min(cur_min, 0) + num;
			min_sum = min(min_sum, cur_min);
			sum += num;
		}
		return sum == min_sum ? max_sum : max(max_sum, sum - min_sum);
	}
};

File: LeetCode/Java/maximumsumcircularsubarray/MaximumSumCircularSubarray.java
package maximumsumcircularsubarray;

public class MaximumSumCircularSubarray {
	public static void main(String[] args) {
		Solution sol = new Solution();
		System.out.println(sol.maxSubarraySumCircular(new int[]{5, -3, 5}));
	}
}

// https://leetcode.com/problems/maximum-sum-circular-subarray/

//IMPORTANT!! Submit Code Region Begin(Do not remove this line)
class Solution {
	public int maxSubarraySumCircular(int[] nums) {
		/*
		 * Kadane's algo:
		 *  1. maxSum: max subarray sum (non-circular)
		 *  2. minSum: min subarray sum
		 *  3. totalSum: sum of all array elements
		 *
		 * Circular case:
		 *  Maximum circular subarray sum = (totalSum - minSum)
		 *  (removes the min subarray to wrap around)
		 *
		 * Edge case:
		 *  If all elements are negative (totalSum == minSum),
		 *      then (totalSum - minSum) would be 0,
		 *      which is invalid (return the maximum possible sum of a non-empty subarray of nums)
		 *  So return maxSum (standard non-circular result),
		 *      Math.max(maxSum, totalSum - minSum)
		 */

		int curMin = nums[0];
		int curMax = nums[0];
		int minSum = nums[0];
		int maxSum = nums[0];
		int totalSum = nums[0];
		int numsLength = nums.length;

		for (int i = 1; i < numsLength; i++) {
			// Normal Kadane
			curMax = Math.max(curMax + nums[i], nums[i]);
			maxSum = Math.max(curMax, maxSum);

			// Kadane with minimum subarray
			curMin = Math.min(curMin + nums[i], nums[i]);
			minSum = Math.min(curMin, minSum);

			totalSum += nums[i];
		}

		if (totalSum == minSum) {
			return maxSum;
		}

		return Math.max((totalSum - minSum), maxSum);
	}
}
//IMPORTANT!! Submit Code Region End(Do not remove this line)